### PR TITLE
Add tests for ProductID metadata consistency for ROOT and streamer formats

### LINKED ISSUE #887: Add the reproducer as a unit test

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR adds the reproducer I mentioned in https://github.com/cms-sw/cmssw/issues/44643#issuecomment-2046131487 as a unit test. It also adds a utility `CatStreamerFiles` to concatenate streamer files such that the InitMessage of only the first file is kept (which is effectively what DAQ does online).

Resolves https://github.com/cms-sw/framework-team/issues/887

#### PR validation:

Added unit tests run

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Eventually to be backported to 14_0_X

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones please review

@wddgit These are the tests we discussed yesterday

File: FWCore/Framework/test/stubs/ToyIntProducers.cc [hunk 1]
```diff
@@ -1016,6 +1039,7 @@ namespace edm::test {
       edm::EDPutTokenT<edmtest::IntProduct> token_;
       int value_;
     };
+    using IntTransformer = IntProducer;
```
Comment (Dr15Jones) [inline FWCore/Framework/test/stubs/ToyIntProducers.cc:1042]: Best to add a comment explaining why this is different.

File: IOPool/Input/test/BuildFile.xml [hunk 2]
```diff
@@ -25,4 +25,7 @@
   <test name="TestIOPoolInputNoParentDictionary" command="testNoParentDictionary.sh"/>
   <test name="TestFileOpenErrorExitCode" command="testFileOpenErrorExitCode.sh"/>
   <test name="TestIOPoolInputSchemaEvolution" command="testSchemaEvolution.sh"/>
+
+  <test name="TestIOPoolInputRefProductIDMetadataConsistency" command="run_TestRefProductIDMetadataConsistencyRoot.sh"/>
```
Comment (Dr15Jones) [inline IOPool/Input/test/BuildFile.xml:29]: Why a .sh instead of having the tests run cmsRun directly?

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

File: FWCore/Framework/test/stubs/ToyIntProducers.cc (hunk 1 above)
Comment (makortel) [reply FWCore/Framework/test/stubs/ToyIntProducers.cc:1047]: Thanks, I added a comment.

File: IOPool/Input/test/BuildFile.xml (hunk 2 above)
Comment (makortel) [reply IOPool/Input/test/BuildFile.xml:29]: Partially out of old habit, but running the four dependent `cmsRun` jobs in a script feels easier than expressing the DAG in the `BuildFile.xml`?

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Comparisons show the usual
- https://github.com/cms-sw/cmssw/issues/43293
- https://github.com/cms-sw/cmssw/issues/43790

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Pull request merged**